In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option("display.max_columns", None)

In [2]:
df = pd.read_csv("../data/bank-full.csv", sep = ";")

print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

df.head()

Rows: 45211
Columns: 17


,age,job,marital,education,default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome,y
0,58,management,married,tertiary,no,2143,yes,no,unknown,5,may,261,1,-1,0,unknown,no
1,44,technician,single,secondary,no,29,yes,no,unknown,5,may,151,1,-1,0,unknown,no
2,33,entrepreneur,married,secondary,no,2,yes,yes,unknown,5,may,76,1,-1,0,unknown,no
3,47,blue-collar,married,unknown,no,1506,yes,no,unknown,5,may,92,1,-1,0,unknown,no
4,33,unknown,single,unknown,no,1,no,no,unknown,5,may,198,1,-1,0,unknown,no


In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 45211 entries, 0 to 45210
Data columns (total 17 columns):
 #   Column     Non-Null Count  Dtype
---  ------     --------------  -----
 0   age        45211 non-null  int64
 1   job        45211 non-null  str  
 2   marital    45211 non-null  str  
 3   education  45211 non-null  str  
 4   default    45211 non-null  str  
 5   balance    45211 non-null  int64
 6   housing    45211 non-null  str  
 7   loan       45211 non-null  str  
 8   contact    45211 non-null  str  
 9   day        45211 non-null  int64
 10  month      45211 non-null  str  
 11  duration   45211 non-null  int64
 12  campaign   45211 non-null  int64
 13  pdays      45211 non-null  int64
 14  previous   45211 non-null  int64
 15  poutcome   45211 non-null  str  
 16  y          45211 non-null  str  
dtypes: int64(7), str(10)
memory usage: 5.9 MB


In [5]:
df.isnull().sum()

age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
y            0
dtype: int64

In [7]:
print("Duplicate Rows:", df.duplicated().sum())

Duplicate Rows: 0


In [8]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
age,45211.0,40.936210,10.618762,18.0,33.0,39.0,48.0,95.0
balance,45211.0,1362.272058,3044.765829,-8019.0,72.0,448.0,1428.0,102127.0
day,45211.0,15.806419,8.322476,1.0,8.0,16.0,21.0,31.0
duration,45211.0,258.163080,257.527812,0.0,103.0,180.0,319.0,4918.0
campaign,45211.0,2.763841,3.098021,1.0,1.0,2.0,3.0,63.0
pdays,45211.0,40.197828,100.128746,-1.0,-1.0,-1.0,-1.0,871.0
previous,45211.0,0.580323,2.303441,0.0,0.0,0.0,0.0,275.0


In [9]:
for col in df.select_dtypes(include="object").columns:
    print("/n", col)
    print(df[col].value_counts())

/n job
job
blue-collar      9732
management       9458
technician       7597
admin.           5171
services         4154
retired          2264
self-employed    1579
entrepreneur     1487
unemployed       1303
housemaid        1240
student           938
unknown           288
Name: count, dtype: int64
/n marital
marital
married     27214
single      12790
divorced     5207
Name: count, dtype: int64
/n education
education
secondary    23202
tertiary     13301
primary       6851
unknown       1857
Name: count, dtype: int64
/n default
default
no     44396
yes      815
Name: count, dtype: int64
/n housing
housing
yes    25130
no     20081
Name: count, dtype: int64
/n loan
loan
no     37967
yes     7244
Name: count, dtype: int64
/n contact
contact
cellular     29285
unknown      13020
telephone     2906
Name: count, dtype: int64
/n month
month
may    13766
jul     6895
aug     6247
jun     5341
nov     3970
apr     2932
feb     2649
jan     1403
oct      738
sep      579
mar      477
dec     

C:\Users\thiru\AppData\Local\Temp\ipykernel_2908\3913055973.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  for col in df.select_dtypes(include="object").columns:


In [10]:
for col in ['job', 'education', 'contact', 'poutcome']:
    print(f"/n{col}")
    print(df[col].value_counts(normalize=True)*100)

/njob
job
blue-collar      21.525735
management       20.919688
technician       16.803433
admin.           11.437482
services          9.188029
retired           5.007631
self-employed     3.492513
entrepreneur      3.289023
unemployed        2.882042
housemaid         2.742695
student           2.074716
unknown           0.637013
Name: proportion, dtype: float64
/neducation
education
secondary    51.319369
tertiary     29.419831
primary      15.153392
unknown       4.107407
Name: proportion, dtype: float64
/ncontact
contact
cellular     64.774059
unknown      28.798301
telephone     6.427639
Name: proportion, dtype: float64
/npoutcome
poutcome
unknown    81.747805
failure    10.840282
other       4.069806
success     3.342107
Name: proportion, dtype: float64


In [11]:
df['age_group'] = pd.cut(
    df['age'],
    bins = [18, 30, 40, 50, 60, 100],
    labels = [
        '18-30',
        '31-40',
        '41-50',
        '51-60',
        '60+'
    ]
)

In [12]:
df['balance_segment'] = pd.cut(
    df['balance'],
    bins=[-10000, 0, 1000, 5000, 100000],
    labels= [
        'Negative',
        'Low',
        'Medium',
        'High'
    ]
)

In [14]:
df['campaign_group'] = pd.cut(
    df['campaign'],
    bins=[0, 2, 5, 10, 100],
    labels=[
        'Low',
        'Medium',
        'High',
        'Very High'
    ]
)

In [15]:
df['subscription'] = np.where(
    df['y'] == 'yes',
    1,
    0
)

In [16]:
df.head()

,age,job,marital,education,default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome,y,age_group,balance_segment,campaign_group,subscription
0,58,management,married,tertiary,no,2143,yes,no,unknown,5,may,261,1,-1,0,unknown,no,51-60,Medium,Low,0
1,44,technician,single,secondary,no,29,yes,no,unknown,5,may,151,1,-1,0,unknown,no,41-50,Low,Low,0
2,33,entrepreneur,married,secondary,no,2,yes,yes,unknown,5,may,76,1,-1,0,unknown,no,31-40,Low,Low,0
3,47,blue-collar,married,unknown,no,1506,yes,no,unknown,5,may,92,1,-1,0,unknown,no,41-50,Medium,Low,0
4,33,unknown,single,unknown,no,1,no,no,unknown,5,may,198,1,-1,0,unknown,no,31-40,Low,Low,0


In [18]:
df.to_csv("../data/bank_cleaned.csv", index= False)

print("Export Complete")

Export Complete


## Data Cleaning Summary

- Dataset contains 45,211 banking customer records.
- No missing values were observed.
- Unknown categories were retained as business-relevant information.
- Age groups, balance segments, and campaign intensity segments were engineered.
- Target variable was converted into a binary subscription indicator.

In [19]:
df[['age_group',
    'balance_segment',
    'campaign_group',
    'subscription']].head()

,age_group,balance_segment,campaign_group,subscription
0,51-60,Medium,Low,0
1,41-50,Low,Low,0
2,31-40,Low,Low,0
3,41-50,Medium,Low,0
4,31-40,Low,Low,0


In [20]:
df.shape

(45211, 21)